dart/common/detail/TemplateJoinerDispatchMacro.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2016, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Michael X. Grey <mxgrey@gatech.edu>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_DETAIL_TEMPLATEJOINERDISPATCHMACRO_H_
+#define DART_COMMON_DETAIL_TEMPLATEJOINERDISPATCHMACRO_H_
+
+//==============================================================================
+/// This macro provides the implementation for most of the member functions in
+/// common::AddonManagerJoiner, dynamics::NodeManagerJoinerForBodyNode, and
+/// NodeManagerJoinerForSkeleton. The member functions of those classes share
+/// essentially the same logic, so it makes sense to have a single macro that
+/// provides the implementation for all of them.
+#define DETAIL_DART_COMMON_TEMPLATEJOINERDISPATCH_IMPL(ReturnType, ClassName, Function, Suffix, Args)\
+  template <class Base1, class Base2>\
+  template <class T>\
+  ReturnType ClassName <Base1, Base2>:: Function Suffix\
+  {\
+    if(Base1::template isSpecializedFor<T>())\
+      return Base1::template Function <T> Args ;\
+  \
+    return Base2::template Function <T> Args ;\
```

Comment:
I do have one last concern, though... in the current scheme, if a user inherits the single-argument ``SpecializedAddonManager`` or ``SpecializedNodeManager`` directly (i.e. non-virtually), then it could cause issues. I'm thinking we should actually split out the iterator into a base class (maybe called something like ``detail::SpecializedAddonManagerData``) which will be virtually inherited by ``SpecializedNodeManager``. I think that would **guarantee** under all circumstances that there aren't any multiple-inheritance conflicts.